File: 6. Minimum Cost Climbing Stairs.py
# 6. Minimum Cost Climbing Stairs
# Topic: Minimum Cost Climbing Stairs
# Description: Given a list of costs for stairs, find the minimal cost to reach the top using recursion + memoization.
# Input: n followed by n costs
# Output: minimum total cost

import sys
from functools import lru_cache

if __name__ == '__main__':
    data = sys.stdin.read().strip().split()
    if not data:
        sys.exit(0)
    it = iter(data)
    n = int(next(it))
    cost = [int(next(it)) for _ in range(n)]

    @lru_cache(maxsize=None)
    def min_cost(i: int) -> int:
        if i >= n:
            return 0
        return cost[i] + min(min_cost(i+1), min_cost(i+2))

    print(min(min_cost(0), min_cost(1)))

# Memoization starts at min_cost decorated with lru_cache.
# Time Complexity: O(n)
# Space Complexity: O(n)

File: 6. Minimum Cost Climbing Stairs.cpp
// 6. Minimum Cost Climbing Stairs
// Topic: Minimum Cost Climbing Stairs
// Description: Given cost[] array, can start at step 0 or 1; find min cost to reach beyond last
// step using memoization. Input: n (number of steps) followed by n costs Output: minimum cost

#include <bits/stdc++.h>
using namespace std;

// Memoization starts here: vector<int> memo
int minCost(int i, const vector<int>& cost, vector<int>& memo)
{
    int n = cost.size();
    if (i >= n)
        return 0;
    if (memo[i] != INT_MAX)
        return memo[i];
    int take1 = cost[i] + minCost(i + 1, cost, memo);
    int take2 = cost[i] + minCost(i + 2, cost, memo);
    memo[i] = min(take1, take2);
    return memo[i];
}

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    int n;
    if (!(cin >> n))
        return 0;
    vector<int> cost(n);
    for (int i = 0; i < n; ++i) cin >> cost[i];
    vector<int> memo(n, INT_MAX);
    int res = min(minCost(0, cost, memo), minCost(1, cost, memo));
    cout << res << '\n';
}

// Memoization process started in minCost where memo vector is used to cache results.
// Time Complexity: O(n)
// Space Complexity: O(n)
